File: examples/python/textual/notes_showcase/src/notes_showcase/viewmodels/status_bar_vm.py
"""StatusBarVM — read-only VM driving the three status-bar slots.

Each slot (``note_count_text``, ``starred_text``, ``editing_text``) is a
:class:`vmx.DerivedProperty` whose source is a :class:`reactivex.BehaviorSubject`
that re-emits the watched VM whenever a relevant ``PropertyChangedMessage``
appears on its hub. Equality-guard inside :class:`DerivedProperty` means
``value_changed`` only fires when the rendered string actually differs.
"""

from __future__ import annotations

import dataclasses
from typing import cast

from reactivex.abc import DisposableBase
from reactivex.subject import BehaviorSubject

from vmx import (
    ComponentVM,
    DerivedProperty,
    MessageHub,
    PropertyChangedMessage,
    RxDispatcher,
    from_sources,
)
from vmx.messages.protocols import Message
from vmx.services.dispatcher import Dispatcher

from notes_showcase.viewmodels.note_form_vm import NoteFormVM
from notes_showcase.viewmodels.notebooks_root_vm import NotebooksRootVM
from notes_showcase.viewmodels.notes_view_vm import NotesViewVM


class StatusBarVM(ComponentVM):
    """Three-slot status bar fed by three source VMs."""

    def __init__(
        self,
        *,
        name: str,
        hint: str,
        hub: MessageHub[Message],
        dispatcher: Dispatcher,
        notes_view: NotesViewVM,
        notebooks: NotebooksRootVM,
        note_form: NoteFormVM,
    ) -> None:
        super().__init__(name=name, hint=hint, hub=hub, dispatcher=dispatcher)
        self._notes_view = notes_view
        self._notebooks = notebooks
        self._note_form = note_form

        self._notes_view_subject: BehaviorSubject[NotesViewVM] = BehaviorSubject(
            notes_view
        )
        self._notebooks_subject: BehaviorSubject[NotebooksRootVM] = BehaviorSubject(
            notebooks
        )
        self._note_form_subject: BehaviorSubject[NoteFormVM] = BehaviorSubject(
            note_form
        )

        def _resub(subject: BehaviorSubject[object], target: object) -> DisposableBase:
            def _on_msg(m: Message) -> None:
                if isinstance(m, PropertyChangedMessage) and m.sender is target:
                    subject.on_next(target)

            return notes_view.hub.messages.subscribe(on_next=_on_msg)

        self._subs: list[DisposableBase] = [
            _resub(
                cast("BehaviorSubject[object]", self._notes_view_subject), notes_view
            ),
            _resub(cast("BehaviorSubject[object]", self._notebooks_subject), notebooks),
            _resub(cast("BehaviorSubject[object]", self._note_form_subject), note_form),
        ]

        self._note_count_text: DerivedProperty[str] = from_sources(
            self._notes_view_subject,
            transform=lambda nv: _render_note_count(cast(NotesViewVM, nv)),
        )
        self._starred_text: DerivedProperty[str] = from_sources(
            self._notes_view_subject,
            transform=lambda nv: _render_starred(cast(NotesViewVM, nv)),
        )
        self._editing_text: DerivedProperty[str] = from_sources(
            self._note_form_subject,
            transform=lambda nf: _render_editing(cast(NoteFormVM, nf)),
        )

    # ── Public surface ─────────────────────────────────────────────────────
    @property
    def hub(self) -> MessageHub[Message]:
        return self._hub

    @property
    def note_count_text(self) -> DerivedProperty[str]:
        return self._note_count_text

    @property
    def starred_text(self) -> DerivedProperty[str]:
        return self._starred_text

    @property
    def editing_text(self) -> DerivedProperty[str]:
        return self._editing_text

    # ── Lifecycle override ─────────────────────────────────────────────────
    def _on_dispose(self) -> None:
        for sub in self._subs:
            sub.dispose()
        self._note_count_text.dispose()
        self._starred_text.dispose()
        self._editing_text.dispose()
        self._notes_view_subject.on_completed()
        self._notes_view_subject.dispose()
        self._notebooks_subject.on_completed()
        self._notebooks_subject.dispose()
        self._note_form_subject.on_completed()
        self._note_form_subject.dispose()
        super()._on_dispose()

    # ── Builder entry-point ────────────────────────────────────────────────
    @staticmethod
    def builder() -> StatusBarVMBuilder:  # type: ignore[override]
        # Narrows ComponentVM.builder() to the showcase StatusBarVMBuilder.
        return StatusBarVMBuilder()


def _render_note_count(nv: NotesViewVM) -> str:
    count = len(nv.filtered_items)
    return f"{count} note{'' if count == 1 else 's'}"


def _render_starred(nv: NotesViewVM) -> str:
    k = sum(1 for n in nv.filtered_items if n.model.starred)
    return f"{k} starred"


def _render_editing(nf: NoteFormVM) -> str:
    if not nf.has_bound_note:
        return "No selection"
    dirty_marker = " *" if nf.is_dirty.value else ""
    return f"Editing: {nf.draft.title}{dirty_marker}"


@dataclasses.dataclass(frozen=True, slots=True)
class StatusBarVMBuilder:
    """Immutable fluent builder for :class:`StatusBarVM`."""

    _name: str | None = None
    _hint: str = ""
    _hub: MessageHub[Message] | None = None
    _dispatcher: Dispatcher | None = None
    _notes_view: NotesViewVM | None = None
    _notebooks: NotebooksRootVM | None = None
    _note_form: NoteFormVM | None = None

    def name(self, value: str) -> StatusBarVMBuilder:
        return dataclasses.replace(self, _name=value)

    def hint(self, value: str) -> StatusBarVMBuilder:
        return dataclasses.replace(self, _hint=value)

    def services(
        self, hub: MessageHub[Message], dispatcher: Dispatcher
    ) -> StatusBarVMBuilder:
        return dataclasses.replace(self, _hub=hub, _dispatcher=dispatcher)

    def notes_view(self, value: NotesViewVM) -> StatusBarVMBuilder:
        return dataclasses.replace(self, _notes_view=value)

    def notebooks(self, value: NotebooksRootVM) -> StatusBarVMBuilder:
        return dataclasses.replace(self, _notebooks=value)

    def note_form(self, value: NoteFormVM) -> StatusBarVMBuilder:
        return dataclasses.replace(self, _note_form=value)

    def build(self) -> StatusBarVM:
        if self._name is None:
            raise ValueError("name is required")
        if self._notes_view is None:
            raise ValueError("notes_view is required")
        if self._notebooks is None:
            raise ValueError("notebooks is required")
        if self._note_form is None:
            raise ValueError("note_form is required")
        hub = self._hub if self._hub is not None else MessageHub[Message]()
        dispatcher = (
            self._dispatcher
            if self._dispatcher is not None
            else RxDispatcher.immediate()
        )
        return StatusBarVM(
            name=self._name,
            hint=self._hint,
            hub=hub,
            dispatcher=dispatcher,
            notes_view=self._notes_view,
            notebooks=self._notebooks,
            note_form=self._note_form,
        )

File: examples/swift/notes-showcase/Sources/NotesShowcaseCore/ViewModels/StatusBarVM.swift
//
// StatusBarVM — three DerivedProperty<String> slots for the status bar.
//
// Ports examples/csharp/avalonia/NotesShowcase/ViewModels/StatusBarVM.cs.
// See task-6-brief.md.
//
// Three slots (all `DerivedProperty<String>`, read-only):
//   noteCountText  — "N note(s)"      from NotesViewVM.filteredItems.count
//   starredText    — "K starred"       from NotesViewVM.filteredItems (starred filter)
//   editingText    — "Editing: TITLE [*]" / "No selection" from NoteFormVM state
//
// Each slot is backed by a `CurrentValueSubject` (= C# BehaviorSubject).
// Hub subscriptions detect property changes from the source VMs and re-send
// the source VM on the subject, triggering a DerivedProperty recompute.
// The re-send is foreground-marshalled to keep recomputes on the UI thread.
//
// Note: BindableDerived sidecars are skipped here — they are wired in Task 8
// (Swift has no INPC, so SwiftUI/AppKit bindings use a different mechanism).
//
// Cross-module subclassing enabled by ADR-0066: `hub`, `dispatcher`, and
// `_raisePropertyChanged` are `public` on `ComponentVMBase`.
//
import Foundation
import Combine
import VMx

/// Read-only VM driving the three status-bar slots.
///
/// Each slot is a `DerivedProperty<String>` recomputed whenever the relevant
/// source VM emits a `PropertyChangedMessage` on the shared hub.
public final class StatusBarVM: ComponentVMBase {

    // ── Private state ──────────────────────────────────────────────────────

    private let _notesView: NotesViewVM
    private let _noteForm: NoteFormVM

    private let _nvSubject: CurrentValueSubject<NotesViewVM, Never>
    private let _nfSubject: CurrentValueSubject<NoteFormVM, Never>

    private var _nvCancellable: AnyCancellable?
    private var _nfCancellable: AnyCancellable?

    // ── Derived properties ─────────────────────────────────────────────────

    /// "N note(s)" — total count of filtered items in the note list.
    public let noteCountText: DerivedProperty<String>

    /// "K starred" — count of starred items in the filtered list.
    public let starredText: DerivedProperty<String>

    /// "Editing: TITLE [*]" when the form is bound and dirty;
    /// "Editing: TITLE" when bound and clean; "No selection" when unbound.
    public let editingText: DerivedProperty<String>

    // ── Init ───────────────────────────────────────────────────────────────

    private init(
        name: String,
        hint: String,
        hub: MessageHubProtocol,
        dispatcher: Dispatcher,
        notesView: NotesViewVM,
        noteForm: NoteFormVM
    ) {
        _notesView = notesView
        _noteForm  = noteForm

        // Phase 1: create subjects (BehaviorSubject equivalents) before
        // super.init, then create DerivedProperties from them.  We use local
        // vars so that the closures below don't capture `self` before init.
        let nvSubject = CurrentValueSubject<NotesViewVM, Never>(notesView)
        let nfSubject = CurrentValueSubject<NoteFormVM, Never>(noteForm)
        _nvSubject = nvSubject
        _nfSubject = nfSubject

        noteCountText = DerivedProperty<String>.from(nvSubject.eraseToAnyPublisher()) { nv in
            let c = nv.filteredItems.count
            return "\(c) note\(c == 1 ? "" : "s")"
        }
        starredText = DerivedProperty<String>.from(nvSubject.eraseToAnyPublisher()) { nv in
            let k = nv.filteredItems.filter { $0.starred }.count
            return "\(k) starred"
        }
        editingText = DerivedProperty<String>.from(nfSubject.eraseToAnyPublisher()) { nf in
            guard nf.hasBoundNote else { return "No selection" }
            let dirty = nf.isDirty ? " *" : ""
            return "Editing: \(nf.draft.title)\(dirty)"
        }

        super.init(name: name, hint: hint, hub: hub, dispatcher: dispatcher)

        // Phase 2: wire hub subscriptions (safe to capture `self` now).
        //
        // Pattern mirrors C# `.Where(m => ReferenceEquals(m.Sender, notesView))`:
        // filter `PropertyChangedMessage`s whose `sender` is the watched VM,
        // then foreground-marshal the subject re-send.
        _nvCancellable = hub.messages
            .compactMap { $0 as? PropertyChangedMessage }
            .filter { [weak notesView] m in m.sender === notesView }
            .sink { [weak self] _ in
                guard let self else { return }
                self.dispatcher.scheduleForeground { [weak self] in
                    guard let self else { return }
                    self._nvSubject.send(self._notesView)
                }
            }

        _nfCancellable = hub.messages
            .compactMap { $0 as? PropertyChangedMessage }
            .filter { [weak noteForm] m in m.sender === noteForm }
            .sink { [weak self] _ in
                guard let self else { return }
                self.dispatcher.scheduleForeground { [weak self] in
                    guard let self else { return }
                    self._nfSubject.send(self._noteForm)
                }
            }
    }

    // ── Lifecycle overrides ────────────────────────────────────────────────

    public override func _onDispose() {
        _nvCancellable?.cancel()
        _nvCancellable = nil
        _nfCancellable?.cancel()
        _nfCancellable = nil
        _nvSubject.send(completion: .finished)
        _nfSubject.send(completion: .finished)
        noteCountText.dispose()
        starredText.dispose()
        editingText.dispose()
        super._onDispose()
    }

    // ── Builder ────────────────────────────────────────────────────────────

    /// Returns a new empty builder.
    public static func builder() -> StatusBarVMBuilder {
        StatusBarVMBuilder()
    }

    /// Immutable fluent builder for `StatusBarVM`.
    ///
    /// Required: `name(_:)`, `services(hub:dispatcher:)`, `notesView(_:)`,
    /// `notebooks(_:)` (retained for cross-flavor parity; unused internally),
    /// `noteForm(_:)`.
    public struct StatusBarVMBuilder {
        private var _name: String?
        private var _hint: String = ""
        private var _hub: MessageHubProtocol?
        private var _dispatcher: Dispatcher?
        private var _notesView: NotesViewVM?
        private var _noteForm: NoteFormVM?

        fileprivate init() {}

        public func name(_ value: String) -> StatusBarVMBuilder {
            var c = self; c._name = value; return c
        }
        public func hint(_ value: String) -> StatusBarVMBuilder {
            var c = self; c._hint = value; return c
        }
        public func services(hub: MessageHubProtocol, dispatcher: Dispatcher) -> StatusBarVMBuilder {
            var c = self; c._hub = hub; c._dispatcher = dispatcher; return c
        }
        public func notesView(_ vm: NotesViewVM) -> StatusBarVMBuilder {
            var c = self; c._notesView = vm; return c
        }
        /// Kept for cross-flavor builder parity; `NotebooksRootVM` is not
        /// currently used in the status-bar computation.
        public func notebooks(_ vm: NotebooksRootVM) -> StatusBarVMBuilder { self }
        public func noteForm(_ vm: NoteFormVM) -> StatusBarVMBuilder {
            var c = self; c._noteForm = vm; return c
        }

        /// Validates required fields and constructs a `StatusBarVM`.
        ///
        /// - Throws: `BuilderValidationError` if any required field is missing.
        public func build() throws -> StatusBarVM {
            guard let name = _name else { throw BuilderValidationError(missingField: "name") }
            guard let hub = _hub else { throw BuilderValidationError(missingField: "hub") }
            guard let dispatcher = _dispatcher else { throw BuilderValidationError(missingField: "dispatcher") }
            guard let notesView = _notesView else { throw BuilderValidationError(missingField: "notesView") }
            guard let noteForm = _noteForm else { throw BuilderValidationError(missingField: "noteForm") }
            return StatusBarVM(
                name: name, hint: _hint,
                hub: hub, dispatcher: dispatcher,
                notesView: notesView, noteForm: noteForm
            )
        }
    }
}
